cdap-data-fabric/src/main/java/co/cask/cdap/data2/sql/RetryingSqlTransactionRunner.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.sql;
+
+import co.cask.cdap.data2.transaction.Transactions;
+import co.cask.cdap.spi.data.StructuredTableAdmin;
+import co.cask.cdap.spi.data.transaction.TransactionException;
+import co.cask.cdap.spi.data.transaction.TransactionRunner;
+import co.cask.cdap.spi.data.transaction.TxRunnable;
+
+import java.util.concurrent.TimeUnit;
+import javax.sql.DataSource;
+
+/**
+ * Retries SQL operations in case they fail due to conflict.
+ * This class is based on {@link Transactions#createTransactionalWithRetry}.
+ */
+public class RetryingSqlTransactionRunner implements TransactionRunner {
+  private static final String TRANSACTION_CONFLICT_SQL_STATE = "40001";
```

Comment:
Is there a document or file we can learn about these state code, better to put a comment about where to see them

Reply:
I got it from this page - https://www.postgresql.org/docs/9.6/transaction-iso.html. Added a comment to that effect